File: T/1-two-sum/main.go
package main

import (
	"fmt"
)

func twoSum(nums []int, target int) []int {
	m := make(map[int]int)
	for i := 0; i < len(nums); i++ {
		v, ok := m[target-nums[i]]
		if ok {
			return []int{v, i}
		} else {
			m[nums[i]] = i
		}
	}
	return []int{}
}

func main() {
	fmt.Println(twoSum([]int{2, 7, 11, 15}, 9))
}

File: T/1-two-sum/main.rs
struct Solution {}

use std::collections::HashMap;
impl Solution {
  pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut m: HashMap<i32, usize> = HashMap::new();

    for (idx, v) in nums.iter().enumerate() {
      let diff = target - v;
      if m.contains_key(&diff) {
        return vec![m[&diff] as i32, idx as i32];
      }
      m.insert(*v, idx);
    }
    vec![]
  }
}
